File: Basis Server/BasisServerTests/Networking/HelloWorldPeerMessageTests.cs
using System.Collections.Concurrent;
using System.Net;
using System.Net.Sockets;
using Basis.HelloWorld;
using BasisServerHandle;
using Xunit;

namespace BasisServerTests;

// ─────────────────────────────────────────────────────────────────────────────
// End-to-end peer messaging over a real server and real UDP sockets.
//
// Unlike the rest of the networking suite — which drives BasisServerHandleEvents
// synchronously through interface fakes — this boots NetworkServer for real on a
// loopback port and joins sixteen BasisHelloClients through the full handshake:
// version check, password, DID challenge/response, and the metadata reply that
// admits the peer. What it proves is the thing no fake can: that sixteen clients
// can hold a full mesh of directed conversations through one server at once.
//
// Runs in the shared-network-statics collection because NetworkServer's state is
// process-wide.
// ─────────────────────────────────────────────────────────────────────────────

/// <summary>
/// Boots one real server for the whole class. Started once because a boot binds sockets, starts
/// the join-broadcast worker and rebuilds the auth identity — none of it worth paying per test.
/// </summary>
public sealed class HelloWorldServerFixture : IDisposable
{
    public const string Password = "hello-world-integration-test";

    public int Port { get; }

    public HelloWorldServerFixture()
    {
        Port = FindFreeUdpPort();

        NetworkServer.StartServer(new Configuration
        {
            SetPort = (ushort)Port,
            Password = Password,
            UseAuth = true,
            UseAuthIdentity = true,
            // Keeps the run from writing config.xml, permissions.xml and the allow/ban lists into
            // the test binary's folder. Everything the test needs lives in memory.
            HasFileSupport = false,
            EnableStatistics = false,
            EnableConsole = false,
            ApiEnabled = false,
            PeerLimit = 64,
        });
    }

    public void Dispose()
    {
        // StopWorker first: it unsubscribes the handlers, which reads NetworkServer.Listener —
        // and StopServer nulls it. Neither touches BasisServerReductionSystemEvents, whose tick
        // loop is started from a static constructor and cannot be restarted once shut down.
        BasisServerHandleEvents.StopWorker();
        NetworkServer.StopServer();
    }

    /// <summary>
    /// Asks the OS for a port nobody is using. Binding to port 0 and reading back what was
    /// assigned is the only way to get one that is genuinely free; a hard-coded port collides with
    /// whatever else is on the build machine, including a second copy of this suite.
    /// </summary>
    private static int FindFreeUdpPort()
    {
        using Socket probe = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
        probe.Bind(new IPEndPoint(IPAddress.Loopback, 0));
        return ((IPEndPoint)probe.LocalEndPoint!).Port;
    }
}

[Collection("BasisServer shared network statics")]
public sealed class HelloWorldPeerMessageTests : IClassFixture<HelloWorldServerFixture>
{
    private const int ClientCount = 16;
    private static readonly TimeSpan JoinTimeout = TimeSpan.FromSeconds(20);
    private static readonly TimeSpan DeliveryTimeout = TimeSpan.FromSeconds(30);

    private readonly HelloWorldServerFixture _server;

    public HelloWorldPeerMessageTests(HelloWorldServerFixture server)
    {
        _server = server;
    }

    /// <summary>
    /// Sixteen clients join, then every one of them sends a distinct directed message to each of
    /// the other fifteen — 240 messages across a full mesh. Every client must receive exactly the
    /// fifteen addressed to it, from the right sender, and nothing addressed to anyone else.
    /// </summary>
    [Fact]
    public void SixteenClients_ExchangeDirectedMessagesAcrossTheFullMesh()
    {
        BasisHelloClient[] clients = JoinClients(ClientCount);
        try
        {
            // Received text per client, tagged with the sender's player id. A bag rather than a
            // list because the fifteen senders arrive concurrently on each client's pump thread.
            var inbox = new ConcurrentBag<(ushort Sender, string Text)>[ClientCount];
            for (int i = 0; i < ClientCount; i++)
            {
                inbox[i] = new ConcurrentBag<(ushort, string)>();
                int index = i;
                clients[i].TextReceived += (sender, text, _) => inbox[index].Add((sender, text));
            }

            for (int from = 0; from < ClientCount; from++)
            {
                for (int to = 0; to < ClientCount; to++)
                {
                    if (from == to) continue;
                    clients[from].SendText(clients[to].PlayerId, MessageFor(from, to));
                }
            }

            WaitUntil(
                () => inbox.All(bag => bag.Count >= ClientCount - 1),
                DeliveryTimeout,
                () => $"only {inbox.Sum(bag => bag.Count)} of {ClientCount * (ClientCount - 1)} messages arrived " +
                      $"(per client: {string.Join(", ", inbox.Select(bag => bag.Count))})");

            for (int to = 0; to < ClientCount; to++)
            {
                (ushort Sender, string Text)[] received = inbox[to].ToArray();

                // Exactly fifteen: an extra would mean the server relayed a message to someone it
                // was not addressed to, which is the failure that matters most here.
                Assert.Equal(ClientCount - 1, received.Length);

                for (int from = 0; from < ClientCount; from++)
                {
                    if (from == to) continue;
                    ushort senderId = clients[from].PlayerId;
                    Assert.Contains((senderId, MessageFor(from, to)), received);
                }
            }

            // The example from the request, spelled out: client 10 (0x0A) to client 15 (0x0F).
            Assert.Contains(
                (clients[10].PlayerId, "hello0A_0F"),
                inbox[15].ToArray());
        }
        finally
        {
            DisconnectAll(clients);
        }
    }

    /// <summary>
    /// The hello-world behaviour itself: a number passed around a ring of sixteen clients, each
    /// one adding 1 and handing it to its neighbour. Sixteen hops means the volley crosses every
    /// client-to-client edge of the ring and comes back to where it started.
    /// </summary>
    [Fact]
    public void SixteenClients_EchoNumbersAroundTheRing()
    {
        BasisHelloClient[] clients = JoinClients(ClientCount);
        try
        {
            const int FinalValue = ClientCount;
            var hops = new ConcurrentQueue<(int Receiver, ushort Sender, int Value)>();
            var finished = new ManualResetEventSlim(false);

            for (int i = 0; i < ClientCount; i++)
            {
                int index = i;
                BasisHelloClient self = clients[i];
                BasisHelloClient next = clients[(i + 1) % ClientCount];

                self.NumberReceived += (sender, value, _) =>
                {
                    hops.Enqueue((index, sender, value));
                    if (value >= FinalValue) finished.Set();
                    else self.SendNumber(next.PlayerId, value + 1);
                };
            }

            clients[0].SendNumber(clients[1].PlayerId, 1);

            WaitUntil(
                () => finished.IsSet,
                DeliveryTimeout,
                () => $"the volley stopped after {hops.Count} hops: " +
                      $"{string.Join(" -> ", hops.Select(h => $"c{h.Receiver}={h.Value}"))}");

            (int Receiver, ushort Sender, int Value)[] ordered = hops.OrderBy(h => h.Value).ToArray();
            Assert.Equal(FinalValue, ordered.Length);

            for (int hop = 0; hop < ordered.Length; hop++)
            {
                int expectedValue = hop + 1;
                int expectedReceiver = (hop + 1) % ClientCount;
                int expectedSender = hop % ClientCount;

                Assert.Equal(expectedValue, ordered[hop].Value);
                Assert.Equal(expectedReceiver, ordered[hop].Receiver);
                Assert.Equal(clients[expectedSender].PlayerId, ordered[hop].Sender);
            }
        }
        finally
        {
            DisconnectAll(clients);
        }
    }

    /// <summary>The message the request asked for: "hello0A_0F" is client 10 talking to client 15.</summary>
    private static string MessageFor(int from, int to) => $"hello{from:X2}_{to:X2}";

    private BasisHelloClient[] JoinClients(int count)
    {
        var clients = new BasisHelloClient[count];
        try
        {
            for (int i = 0; i < count; i++)
            {
                clients[i] = new BasisHelloClient($"Hello{i:X2}");
                Assert.True(
                    clients[i].Connect(IPAddress.Loopback.ToString(), _server.Port, HelloWorldServerFixture.Password, JoinTimeout),
                    $"client {i} did not join the server on port {_server.Port} within {JoinTimeout.TotalSeconds}s");
            }
        }
        catch
        {
            DisconnectAll(clients);
            throw;
        }

        // Distinct ids are what makes the addressing above meaningful — two clients sharing one id
        // would let a mesh test pass while every message went to the wrong peer.
        ushort[] ids = clients.Select(c => c.PlayerId).ToArray();
        Assert.Equal(count, ids.Distinct().Count());

        return clients;
    }

    private static void DisconnectAll(BasisHelloClient?[] clients)
    {
        foreach (BasisHelloClient? client in clients)
        {
            // Teardown must never mask the failure that brought us here: a client that could not
            // finish connecting may also not be able to finish disconnecting.
            try { client?.Dispose(); }
            catch (Exception ex) { Console.WriteLine($"cleanup of a hello client failed: {ex}"); }
        }
    }

    private static void WaitUntil(Func<bool> condition, TimeSpan timeout, Func<string> describeFailure)
    {
        DateTime deadline = DateTime.UtcNow + timeout;
        while (DateTime.UtcNow < deadline)
        {
            if (condition()) return;
            Thread.Sleep(25);
        }

        Assert.Fail($"Timed out after {timeout.TotalSeconds}s: {describeFailure()}");
    }
}

File: basis_server/basis_server_tests/tests/networking/hello_world_peer_message_tests.rs
//! Port of `HelloWorldPeerMessageTests.cs`.
//!
//! End-to-end peer messaging over a real server and real sockets: sixteen hello clients join
//! through the full handshake — version check, password, DID challenge/response, and the
//! metadata reply that admits the peer — and hold a full mesh of directed conversations.

use std::sync::{Arc, Mutex};
use std::time::Duration;

use basis_hello_world_client::BasisHelloClient;
use basis_server_tests::support::{HelloWorldServerFixture, wait_until};
use serial_test::serial;

const CLIENT_COUNT: usize = 16;
const JOIN_TIMEOUT: Duration = Duration::from_secs(20);
const DELIVERY_TIMEOUT: Duration = Duration::from_secs(30);

/// The message the request asked for: "hello0A_0F" is client 10 talking to client 15.
fn message_for(from: usize, to: usize) -> String {
    format!("hello{from:02X}_{to:02X}")
}

fn join_clients(server: &HelloWorldServerFixture, count: usize) -> Vec<Arc<BasisHelloClient>> {
    let mut clients = Vec::with_capacity(count);
    for i in 0..count {
        let client = BasisHelloClient::new(&format!("Hello{i:02X}")).unwrap_or_else(|e| panic!("client {i}: {}", e.report()));
        let joined = client.connect(server.connection_string(), 0, HelloWorldServerFixture::PASSWORD, JOIN_TIMEOUT).unwrap_or_else(|e| panic!("client {i} could not connect: {}", e.report()));
        assert!(joined, "client {i} did not join the server at {} within {}s", server.connection_string(), JOIN_TIMEOUT.as_secs());
        clients.push(client);
    }
    // Distinct ids are what makes the addressing meaningful — two clients sharing one id would
    // let a mesh test pass while every message went to the wrong peer.
    let mut ids: Vec<u16> = clients.iter().map(|c| c.player_id()).collect();
    ids.sort_unstable();
    ids.dedup();
    assert_eq!(ids.len(), count, "player ids must be distinct");
    clients
}

fn disconnect_all(clients: &[Arc<BasisHelloClient>]) {
    for client in clients {
        client.disconnect();
    }
}

/// Sixteen clients join, then every one of them sends a distinct directed message to each of
/// the other fifteen — 240 messages across a full mesh. Every client must receive exactly the
/// fifteen addressed to it, from the right sender, and nothing addressed to anyone else.
#[test]
#[serial]
fn sixteen_clients_exchange_directed_messages_across_the_full_mesh() {
    let server = HelloWorldServerFixture::new();
    let clients = join_clients(&server, CLIENT_COUNT);

    let inbox: Vec<Arc<Mutex<Vec<(u16, String)>>>> = (0..CLIENT_COUNT).map(|_| Arc::new(Mutex::new(Vec::new()))).collect();
    for (i, client) in clients.iter().enumerate() {
        let bag = inbox[i].clone();
        client.on_text_received(Arc::new(move |sender, text, _| bag.lock().unwrap().push((sender, text))));
    }

    for from in 0..CLIENT_COUNT {
        for to in 0..CLIENT_COUNT {
            if from == to {
                continue;
            }
            clients[from].send_text(clients[to].player_id(), &message_for(from, to)).unwrap_or_else(|e| panic!("{}", e.report()));
        }
    }

    wait_until(
        || inbox.iter().all(|bag| bag.lock().unwrap().len() >= CLIENT_COUNT - 1),
        DELIVERY_TIMEOUT,
        || {
            let counts: Vec<String> = inbox.iter().map(|bag| bag.lock().unwrap().len().to_string()).collect();
            let total: usize = inbox.iter().map(|bag| bag.lock().unwrap().len()).sum();
            format!("only {total} of {} messages arrived (per client: {})", CLIENT_COUNT * (CLIENT_COUNT - 1), counts.join(", "))
        },
    );

    for to in 0..CLIENT_COUNT {
        let received = inbox[to].lock().unwrap().clone();
        // Exactly fifteen: an extra would mean the server relayed a message to someone it was
        // not addressed to, which is the failure that matters most here.
        assert_eq!(received.len(), CLIENT_COUNT - 1, "client {to} received {:?}", received);
        for from in 0..CLIENT_COUNT {
            if from == to {
                continue;
            }
            let expected = (clients[from].player_id(), message_for(from, to));
            assert!(received.contains(&expected), "client {to} did not get {expected:?}");
        }
    }

    // The example from the request, spelled out: client 10 (0x0A) to client 15 (0x0F).
    assert!(inbox[15].lock().unwrap().contains(&(clients[10].player_id(), "hello0A_0F".to_string())));

    disconnect_all(&clients);
}

/// The hello-world behaviour itself: a number passed around a ring of sixteen clients, each one
/// adding 1 and handing it to its neighbour. Sixteen hops means the volley crosses every
/// client-to-client edge of the ring and comes back to where it started.
#[test]
#[serial]
fn sixteen_clients_echo_numbers_around_the_ring() {
    let server = HelloWorldServerFixture::new();
    let clients = join_clients(&server, CLIENT_COUNT);

    const FINAL_VALUE: i32 = CLIENT_COUNT as i32;
    let hops: Arc<Mutex<Vec<(usize, u16, i32)>>> = Arc::new(Mutex::new(Vec::new()));
    let finished = Arc::new(std::sync::atomic::AtomicBool::new(false));

    for i in 0..CLIENT_COUNT {
        let this = clients[i].clone();
        let next = clients[(i + 1) % CLIENT_COUNT].clone();
        let hops = hops.clone();
        let finished = finished.clone();
        clients[i].on_number_received(Arc::new(move |sender, value, _| {
            hops.lock().unwrap().push((i, sender, value));
            if value >= FINAL_VALUE {
                finished.store(true, std::sync::atomic::Ordering::SeqCst);
            } else if let Err(e) = this.send_number(next.player_id(), value + 1) {
                eprintln!("hop {value} could not be passed on: {}", e.report());
            }
        }));
    }

    clients[0].send_number(clients[1].player_id(), 1).unwrap_or_else(|e| panic!("{}", e.report()));

    wait_until(
        || finished.load(std::sync::atomic::Ordering::SeqCst),
        DELIVERY_TIMEOUT,
        || {
            let hops = hops.lock().unwrap();
            let path: Vec<String> = hops.iter().map(|(r, _, v)| format!("c{r}={v}")).collect();
            format!("the volley stopped after {} hops: {}", hops.len(), path.join(" -> "))
        },
    );

    let mut ordered = hops.lock().unwrap().clone();
    ordered.sort_by_key(|(_, _, value)| *value);
    assert_eq!(ordered.len(), FINAL_VALUE as usize);
    for (hop, (receiver, sender, value)) in ordered.iter().enumerate() {
        assert_eq!(*value, hop as i32 + 1);
        assert_eq!(*receiver, (hop + 1) % CLIENT_COUNT);
        assert_eq!(*sender, clients[hop % CLIENT_COUNT].player_id());
    }

    disconnect_all(&clients);
}

/// Faults the C# suite left to the transport: a wrong password is refused rather than hung on,
/// and a client that never joined cannot send.
#[test]
#[serial]
fn a_wrong_password_is_refused_and_an_unjoined_client_cannot_send() {
    let server = HelloWorldServerFixture::new();
    let client = BasisHelloClient::new("Impostor").unwrap();
    let joined = client.connect(server.connection_string(), 0, "not-the-password", Duration::from_secs(5)).unwrap_or_else(|e| panic!("{}", e.report()));
    assert!(!joined, "a wrong password must not be admitted");
    let err = client.send_number(1, 1).expect_err("sending before joining must fail");
    assert!(err.report().contains("has not joined"), "{}", err.report());
    let again = client.connect(server.connection_string(), 0, HelloWorldServerFixture::PASSWORD, Duration::from_secs(1));
    assert!(again.is_err(), "a client connects once; the second attempt is refused");
    client.disconnect();
}
